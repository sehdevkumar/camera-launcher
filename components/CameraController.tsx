import { CameraView, CameraType, useCameraPermissions } from 'expo-camera';
import { FlashMode } from 'expo-camera/build/legacy/Camera.types';
import { useRef, useState } from 'react';
import { Button, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import * as MediaLibrary from 'expo-media-library';
export default function CameraPage() {
    const [facing, setFacing] = useState<CameraType>('back');
    const [permission, requestPermission] = useCameraPermissions();
    const [permissionMediaResponse, requestMediaPermission] = MediaLibrary.usePermissions();
    const cameraRef = useRef<CameraView>(null);

    if (!permission) {
        // Camera permissions are still loading.
        return <View />;
    }

    if (!permission.granted) {
        // Camera permissions are not granted yet.
        return (
            <View style={styles.container}>
                <Text style={styles.message}>We need your permission to show the camera</Text>
                <Button onPress={requestPermission} title="grant permission" />
            </View>
        );
    }
    if (!permissionMediaResponse?.granted) {
        // Camera permissions are not granted yet.
        return (
            <View style={styles.container}>
                <Text style={styles.message}>We need your permission to show the Media</Text>
                <Button onPress={requestMediaPermission} title="grant permission" />
            </View>
        );
    }


    function toggleCameraFacing() {
        setFacing(current => (current === 'back' ? 'front' : 'back'));

    }

    async function ImagePost(result:any) {
        let localUri = result.uri;
        let filename = localUri.split('/').pop();

        // Infer the type of the image
        let match = /\.(\w+)$/.exec(filename);
        let type = match ? `image/${match[1]}` : `image`;

        // Upload the image using the fetch and FormData APIs
        let formData = new FormData() as any;
        // Assume "photo" is the name of the form field the server expects
        formData?.append('photo', { uri: localUri, name: filename, type });

        console.log(formData,"form data");

        return await fetch('localhost', {
            method: 'POST',
            body: formData,
            headers: {
                'content-type': 'multipart/form-data',
            },
        });
    }

    function OnStartCapture(e:any) {
        if (cameraRef.current) {
            cameraRef?.current
                ?.takePictureAsync({
                    skipProcessing: false,
                })
                .then(async (photoData: any) => {
                    
                    try {
                        const asset = await MediaLibrary.createAssetAsync(photoData.uri);
                        console.log(photoData.uri, asset);
                                 
                        ImagePost(photoData).then(r=> {
                            console.log(photoData);
                        }).catch(err=> {
                             console.log(err)
                        })

                        
                    } catch (error) {
                       console.debug(error) 
                    }

                });
        }
    }

    return (
        <View style={styles.container}>
            <CameraView style={styles.camera} ref={cameraRef} facing={facing} flash={FlashMode.on}>
                <View style={styles.buttonContainer}>
                    <TouchableOpacity  style={styles.button} onPress={toggleCameraFacing}>
                        <Text style={styles.text}>Flip Camera</Text>
                        
                    </TouchableOpacity>
                    <TouchableOpacity style={styles.button} onPress={OnStartCapture}>
                        <Text style={styles.text}>Take Capture</Text>
                    </TouchableOpacity>
                </View>
            </CameraView>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
    },
    message: {
        textAlign: 'center',
        paddingBottom: 10,
    },
    camera: {
        flex: 1,
    },
    buttonContainer: {
        flex: 1,
        flexDirection: 'row',
        backgroundColor: 'transparent',
        margin: 64,
    },
    button: {
        flex: 1,
        alignSelf: 'flex-end',
        alignItems: 'center',
    },
    text: {
        fontSize: 24,
        fontWeight: 'bold',
        color: 'white',
    },
});
