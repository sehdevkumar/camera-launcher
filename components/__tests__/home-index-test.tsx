import HomeScreen from "@/app/(home)";
import { ThemedText } from "@/components/ThemedText";
import * as React from "react";
import renderer from "react-test-renderer";


describe("Testing the Home Index file",()=> {
  
  it(`Home renders correctly`, () => {
    const tree = renderer
      .create(<HomeScreen></HomeScreen>)
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

 

})
