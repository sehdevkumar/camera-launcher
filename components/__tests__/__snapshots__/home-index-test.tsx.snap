// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Home Index file Home renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f0f0f0",
      "flex": 1,
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Image
        source={1}
      />
      <Text
        style={
          {
            "fontSize": 30,
          }
        }
      >
        Select From
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "padding": 10,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#f9f9f9",
              "borderRadius": 10,
              "elevation": 5,
              "flex": 1,
              "justifyContent": "center",
              "margin": 5,
              "minHeight": 100,
              "minWidth": 150,
              "opacity": 1,
              "padding": 20,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#333",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Gate In
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#f9f9f9",
              "borderRadius": 10,
              "elevation": 5,
              "flex": 1,
              "justifyContent": "center",
              "margin": 5,
              "minHeight": 100,
              "minWidth": 150,
              "opacity": 1,
              "padding": 20,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#333",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Gate Out
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 1`] = `
<Text
  style={
    [
      {
        "color": "#11181C",
      },
      {
        "fontSize": 16,
        "lineHeight": 24,
      },
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
>
  Snapshot test!
</Text>
`;
